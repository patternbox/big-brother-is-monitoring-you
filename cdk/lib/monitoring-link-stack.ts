import { Construct } from 'constructs'
import { Stack, StackProps, CfnParameter, Arn, ArnFormat } from 'aws-cdk-lib'
import * as oam from 'aws-cdk-lib/aws-oam'
import * as iam from 'aws-cdk-lib/aws-iam'

export class MonitoringLinkStack extends Stack {

  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props)

    const monitoringSinkArnParam = new CfnParameter(this, 'MonitoringSinkArn', {
      type: 'String',
      description: 'The monitoring sink ARN',
    })

    const accountLabelParam = new CfnParameter(this, 'AccountLabel', {
      type: 'String',
      description: 'The account label',
      default: '$AccountName',
    })

    new oam.CfnLink(this, 'MonitoringLink', {
      resourceTypes: ['AWS::CloudWatch::Metric', 'AWS::Logs::LogGroup'],
      sinkIdentifier: monitoringSinkArnParam.valueAsString,
      labelTemplate: accountLabelParam.valueAsString,
    })

    const sinkArnSplit = Arn.split(monitoringSinkArnParam.valueAsString, ArnFormat.SLASH_RESOURCE_NAME)

    new iam.Role(this, 'CWCrossAccountSharingRole', {
      roleName: 'CloudWatch-CrossAccountSharingRole',
      assumedBy: new iam.AccountPrincipal(sinkArnSplit.account),
      path: '/',
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchReadOnlyAccess'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAutomaticDashboardsAccess'),
      ]
    })    

  }
}
