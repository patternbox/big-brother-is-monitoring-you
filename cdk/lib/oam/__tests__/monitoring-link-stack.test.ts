import { App } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { MonitoringLinkStack } from '../monitoring-link-stack'

describe('MonitoringLinkStack', () => {

  it('should match snapshot', () => {
    const app = new App()

    const stack = new MonitoringLinkStack(app, 'MonitoringLinkStack', {})

    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot()
  })

})
