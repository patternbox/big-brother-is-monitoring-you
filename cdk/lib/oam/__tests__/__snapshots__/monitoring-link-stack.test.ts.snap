// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitoringLinkStack should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "AccountLabel": Object {
      "Default": "$AccountName",
      "Description": "The account label",
      "Type": "String",
    },
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "MonitoringSinkArn": Object {
      "Description": "The monitoring sink ARN",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CWCrossAccountSharingRole54960778": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Fn::Select": Array [
                          4,
                          Object {
                            "Fn::Split": Array [
                              ":",
                              Object {
                                "Ref": "MonitoringSinkArn",
                              },
                            ],
                          },
                        ],
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchReadOnlyAccess",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchAutomaticDashboardsAccess",
              ],
            ],
          },
        ],
        "Path": "/",
        "RoleName": "CloudWatch-CrossAccountSharingRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "MonitoringLink": Object {
      "Properties": Object {
        "LabelTemplate": Object {
          "Ref": "AccountLabel",
        },
        "ResourceTypes": Array [
          "AWS::CloudWatch::Metric",
          "AWS::Logs::LogGroup",
        ],
        "SinkIdentifier": Object {
          "Ref": "MonitoringSinkArn",
        },
      },
      "Type": "AWS::Oam::Link",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
